service/controller/appcatalog/v1/resource/index/current_test.go

```diff
@@ -0,0 +1,127 @@
+
+	"github.com/giantswarm/apiextensions/pkg/apis/application/v1alpha1"
+	"github.com/giantswarm/micrologger/microloggertest"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+func Test_Resource_GetCurrentState(t *testing.T) {
+	tests := []struct {
+		name              string
+		obj               interface{}
+		expectedConfigMap *corev1.ConfigMap
+		errorMatcher      func(error) bool
+	}{
+		{
+			name: "case 0: index configMap already created",
+			obj: &v1alpha1.AppCatalog{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "giantswarm",
+					Namespace: "default",
+					Labels: map[string]string{
+						"app-operator.giantswarm.io/version": "1.0.0",
+					},
+				},
+				Spec: v1alpha1.AppCatalogSpec{
+					Title:       "Giant Swarm",
+					Description: "Catalog of Apps by Giant Swarm",
+					Storage: v1alpha1.AppCatalogSpecStorage{
+						Type: "helm",
+						URL:  "https://giantswarm.github.com/app-catalog/",
+					},
+					LogoURL: "https://s.giantswarm.io/...",
+				},
+			},
+			expectedConfigMap: &corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "giantswarm-index",
+					Namespace: "giantswarm",
+					Labels: map[string]string{
+						"giantswarm.io/managed-by": "app-operator",
+					},
+				},
+				Data: map[string]string{
+					"index.yaml": "test yaml",
+				},
+			},
+		},
+		{
+			name: "case 1: index configMap not found",
+			obj: &v1alpha1.AppCatalog{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "giantswarm",
+					Namespace: "default",
+					Labels: map[string]string{
+						"app-operator.giantswarm.io/version": "1.0.0",
+					},
+				},
+				Spec: v1alpha1.AppCatalogSpec{
+					Title:       "Giant Swarm",
+					Description: "Catalog of Apps by Giant Swarm",
+					Storage: v1alpha1.AppCatalogSpecStorage{
+						Type: "helm",
+						URL:  "https://giantswarm.github.com/app-catalog/",
+					},
+					LogoURL: "https://s.giantswarm.io/...",
+				},
+			},
+			expectedConfigMap: nil,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			objs := make([]runtime.Object, 0, 0)
+			if tc.expectedConfigMap != nil {
+				objs = append(objs, tc.expectedConfigMap)
+			}
+
+			k8sClient := fake.NewSimpleClientset(objs...)
+
+			c := Config{
+				K8sClient: k8sClient,
+				Logger:    microloggertest.New(),
+
+				IndexNamespace: "giantswarm",
+				ProjectName:    "app-operator",
+			}
+			r, err := New(c)
+			if err != nil {
+				t.Fatalf("error == %#v, want nil", err)
+			}
+
+			result, err := r.GetCurrentState(context.Background(), tc.obj)
+			switch {
+			case err != nil && tc.errorMatcher == nil:
+				t.Fatalf("error == %#v, want nil", err)
+			case err == nil && tc.errorMatcher != nil:
+				t.Fatalf("error == nil, want non-nil")
+			case err != nil && !tc.errorMatcher(err):
+				t.Fatalf("error == %#v, want matching", err)
+			}
+
+			if tc.expectedConfigMap == nil {
```

Comment:
Nitpick but I think the conditional logic could be simplified.

```go
if tc.expectedConfigMap == nil && result != nil {
   t.Fatalf("expected nil ConfigMap got %#v", result)
}
```

Reply:
👍 Thanks

Reply:
When both `result`, `expectedConfigMap` is nil, it will run all the code after these, then it would have null pointer reference problem eventually. So I think `else` statement like this is needed.

Reply:
I would just add another test case for the reverse. Style wise I prefer it if the nil handling is checked first and then the expected state is checked. IMO the code is harder to read with the else statements.


```
if tc.expectedConfigMap != nil && result == nil {
   t.Fatal("expected non-nil ConfigMap got nil")
}
```

Reply:
It's not just reverse checking is needed, if `result` is `nil` then we should not convert it to `ConfigMap`. I'm not sure how to avoid it just using if statement like above.

Reply:
The test is using `t.Fatalf` rather than `t.Errorf`. So the test will fail and stop executing if the result is nil when it shouldn't be.

So you just need to call `toConfigMap` after the nil check.

Reply:
So we decide to move on as it is, we can fix that on later with more debugging.